### PR TITLE
Feature: Implement auth microservice authorization

### PR DESCRIPTION
**Pull request by CarlosQ96**
Related to https://github.com/Giveth/impact-graph/issues/515

And the tasks Amin created https://github.com/Giveth/impact-graph/issues/514

The test wallets from mainnet are taken from the @ethersproject/testcases  -> accounts.json.gz file
which contains tons and tons of testing wallets.

### CODE REVIEW

File: src/server/bootstrap.ts [hunk 1]
```diff
@@ -80,55 +81,27 @@ export async function bootstrap() {
...
+            user = await authorizationHandler(authVersion, token);
             req.user = user;
           }
-
-          const userWalletAddress = headers['wallet-address'];
-          if (userWalletAddress) {
-            req.userwalletAddress = userWalletAddress;
-          }
         } catch (error) {
-          // console.error(
```
Comment (CarlosQ96) [inline src/server/bootstrap.ts:121]: Unused comment.

File: src/server/bootstrap.ts [hunk 2]
```diff
@@ -80,55 +81,27 @@ export async function bootstrap() {
...
           if (headers.authorization) {
             token = headers.authorization.split(' ')[1].toString();
-            const secret = config.get('JWT_SECRET') as string;
-
-            const decodedJwt: any = jwt.verify(token, secret);
-
             let user;
-            if (decodedJwt.nextAuth) {
-              user = {
-                email: decodedJwt?.nextauth?.user?.email,
```
Comment (CarlosQ96) [inline src/server/bootstrap.ts:100]: removed this nextauth, honestly we do not use any user object without ID. I don't know what the nextauth is.

File: src/services/authorizationServices.ts [hunk 3]
```diff
@@ -29,3 +36,80 @@ export const getLoggedInUser = async (ctx: MyContext) => {
...
+      user = await validateImpactGraphJwt(token);
+      break;
+    case '2':
+      user = await validateAuthMicroserviceJwt(token);
+      break;
+    default:
+      throw new Error(errorMessages.INVALID_AUTHORIZATION_VERSION);
+  }
+  return user;
+};
```
Comment (CarlosQ96) [inline src/services/authorizationServices.ts:63]: So to handle backward compatibility I added versions. So version 1 will be our current one and the version 2 is the auth microservice. After we fully implement both I can remove the first one. 
This is to not break anything Ramin or Cherik might be doing. @mateodaza suggested.

File: src/repositories/userRepository.ts [hunk 4]
```diff
@@ -22,3 +22,13 @@ export const findUserByWalletAddress = async (
 export const findUserById = (userId: number): Promise<User | undefined> => {
   return User.findOne({ id: userId });
 };
+
+export const createUserWithPublicAddress = async (
```
Comment (aminlatifi) [inline src/repositories/userRepository.ts:29]: This function should include `analytics.identifyUser` too

File: package.json [hunk 5]
```diff
@@ -41,7 +42,7 @@
     "cors": "^2.8.5",
     "dotenv": "^8.2.0",
     "eth-sig-util": "^3.0.1",
-    "ethers": "^5.5.4",
+    "ethers": "5.5.4",
```
Comment (aminlatifi) [inline package.json]: Is this change by intention?

Comment (aminlatifi) [inline package.json]: Is this change by intention?

File: src/services/authorizationService.test.ts [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+      lastName: 'lastName',
+      email: `${new Date().getTime()}-giveth@giveth.com`,
+      url: 'website url',
+      loginType: 'wallet',
+      walletAddress: publicKey,
+    };
+    const user = await User.create(userData).save();
+    const nonceRoute = config.get('AUTH_MICROSERVICE_NONCE_URL') as string;
+    const nonceResult = await Axios.get(nonceRoute);
+    const wallet = new ethers.Wallet(privateKey, provider);
```
Comment (aminlatifi) [inline src/services/authorizationService.test.ts]: AFAIK, provider is optional and is not needed here.
```suggestion
    const wallet = new ethers.Wallet(privateKey);
```

Comment (aminlatifi) [inline src/services/authorizationService.test.ts]: AFAIK, the second argument (provider) is optional and is not needed here.
```suggestion
    const wallet = new ethers.Wallet(privateKey);
```

File: src/services/authorizationService.test.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+    const accessToken = authenticationResult.data.jwt;
+    const jwtUser = await authorizationHandler('2', accessToken);
+    assert.equal(jwtUser.userId, user.id);
+  });
+  it('should decode jwt and create user if it is nonexistent', async () => {
+    const privateKey = process.env.PRIVATE_ETHERS_SECONDARY_TEST_KEY as string;
+    const publicKey = process.env.PUBLIC_ETHERS_SECONDARY_TEST_KEY as string;
+    const nonceRoute = config.get('AUTH_MICROSERVICE_NONCE_URL') as string;
+    const nonceResult = await Axios.get(nonceRoute);
+    const wallet = new ethers.Wallet(privateKey, provider);
```
Comment (aminlatifi) [inline src/services/authorizationService.test.ts]: Like above
```suggestion
    const wallet = new ethers.Wallet(privateKey);
```

Comment (aminlatifi) [inline src/services/authorizationService.test.ts]: Like above
```suggestion
    const wallet = new ethers.Wallet(privateKey);
```

**Review by aminlatifi — CHANGES_REQUESTED** [review]
Greate @CarlosQ96 , just few little things to fix.

File: package.json (hunk 5 above)
Comment (CarlosQ96) [reply package.json]: Reverting this, seems I had an error but I can't reproduce it anymore, maybe it was my node version.

Comment (CarlosQ96) [reply package.json]: Fixed.

**Pull request merged**